classes/cargo.bbclass

```diff
@@ -116,12 +120,23 @@ cargo_do_compile() {
     export TARGET_CXX="${WRAPPER_DIR}/cxx-wrapper.sh"
     export CC="${WRAPPER_DIR}/cc-native-wrapper.sh"
     export CXX="${WRAPPER_DIR}/cxx-native-wrapper.sh"
-    export TARGET_LD="${WRAPPER_DIR}/ld-wrapper.sh"
-    export LD="${WRAPPER_DIR}/ld-native-wrapper.sh"
     export PKG_CONFIG_ALLOW_CROSS="1"
     export LDFLAGS=""
     export RUSTFLAGS="${RUSTFLAGS}"
     export SSH_AUTH_SOCK="${SSH_AUTH_SOCK}"
+    # When RUST_BUILD == RUST_TARGET, we need to use an unstable Rust feature
+    # to specify different build options for the target and the Host.
+    # The Host configuration is set in do_configure() above.
+    if [ "${RUST_BUILD}" = "${RUST_TARGET}" ]; then
+        export __CARGO_TEST_CHANNEL_OVERRIDE_DO_NOT_USE_THIS="nightly"
+        export CARGO_UNSTABLE_TARGET_APPLIES_TO_HOST="true"
+        export CARGO_UNSTABLE_HOST_CONFIG="true"
```

Comment:
Why is this being set? I don't see any `CARGO_HOST_` variables being set here. In buildroot we enable this so we can set the host config option [`CARGO_HOST_RUSTFLAGS`](https://github.com/buildroot/buildroot/blob/2023.08.1/package/pkg-cargo.mk#L53).

Reply:
If I understood correctly, it's because of the setting is in ` ${CARGO_HOME}/config` 
https://github.com/rust-embedded/meta-rust-bin/blob/ebaab20302b586835ed1e113225003fab4cd415b/classes/cargo.bbclass#L68-L70

Reply:
Why not just set `CARGO_HOST_LINKER="${WRAPPER_DIR}/linker-native-wrapper.sh"` instead of using the cargo config file?

You should also be able to get rid of the cargo config entirely by configuring the target linker and any other config file variables something like this as well:
```
CARGO_TARGET_${RUST_TARGET}_LINKER="${WRAPPER_DIR}/linker-wrapper.sh"
```

In general I think sticking to configuring cargo using env variables only seems to be less likely to break than mixing env based configs with config file based configs.

Reply:
> Why is this being set?
As pointed out by @banditopazzo it's for the host configuration.

> Why not just set CARGO_HOST_LINKER="${WRAPPER_DIR}/linker-native-wrapper.sh" instead of using the cargo config file?

When HOST != TARGET we configure both compilers via the `.cargo/config` file. I chose to keep doing this this way to stay as close as possible to what we have today.

I understand you would prefer to completely get rid of `.cargo/config` and move everything to env vars. Sounds good to me too.

Reply:
> When HOST != TARGET we configure both compilers via the `.cargo/config` file. I chose to keep doing this this way to stay as close as possible to what we have today.

In general there should be no need to configure cargo differently for HOST != TARGET vs HOST == TARGET builds if the host/target config splitting env variables are set up correctly AFAIU.

Reply:
> In general there should be no need to configure cargo differently for HOST != TARGET vs HOST == TARGET builds if the host/target config splitting env variables are set up correctly AFAIU.

We still need a way to pass the HOST linker (`linker-native-wrapper.sh`). If we just set the host/target config splitting env variables, then cargo reverts to `gcc`/`ld` with no options when compiling for the host. That might work sometimes but I don't think it's correct because I believe we want to use the sysroot-native libraries when building anything for the host.

On my system, `linker-native-wrapper.sh` contains: 
```
#!/bin/sh
gcc  -L/home/thomas/yocto/build/tmp/work/cortexa57-poky-linux/xxx/1.0-r0/recipe-sysroot-native/usr/lib                         -L/home/thomas/yocto/build/tmp/work/cortexa57-poky-linux/xxx/1.0-r0/recipe-sysroot-native/lib                         -Wl,--enable-new-dtags                         -Wl,-rpath-link,/home/thomas/yocto/build/tmp/work/cortexa57-poky-linux/xxx/1.0-r0/recipe-sysroot-native/usr/lib                         -Wl,-rpath-link,/home/thomas/yocto/build/tmp/work/cortexa57-poky-linux/xxx/1.0-r0/recipe-sysroot-native/lib                         -Wl,-rpath,/home/thomas/yocto/build/tmp/work/cortexa57-poky-linux/xxx/1.0-r0/recipe-sysroot-native/usr/lib                         -Wl,-rpath,/home/thomas/yocto/build/tmp/work/cortexa57-poky-linux/xxx/1.0-r0/recipe-sysroot-native/lib                         -Wl,-O1 -Wl,--allow-shlib-undefined -Wl,--dynamic-linker=/home/thomas/yocto/build/tmp/sysroots-uninative/x86_64-linux/lib/ld-linux-x86-64.so.2 "$@"
```

If we do not set any host option, we will revert to just calling `gcc` and we do not get the `-Wl,-rpath-link,...` setting when building the build-script. Without this setting, when we run the build-script later, it cannot find the right version of glibc. This was causing #83.


There are two ways to set the Host Linker (via `cargo_home/config` like I have done here) or via `CARGO_HOST_LINKER ` ([like in the other PR](https://github.com/rust-embedded/meta-rust-bin/pull/133/files#diff-73769e78fdb283805d09e223ae2ad5444e8c7128ad5a23e9d52b34eba6679009R133)). I chose the former because this way, the config file looks the same whether `TARGET==HOST` or not.

Reply:
> Without this setting, when we run the build-script later, it cannot find the right version of glibc. This was causing #83.

I think that was a different issue where cargo is picking up target linker flags for host builds rather than not picking up additional native flags for the host build, although I'm more familiar with buildroot which does things somewhat differently.

> I chose the former because this way, the config file looks the same whether `TARGET==HOST` or not.

I mean, why not just get rid of the cargo config file entirely and pass all config options as env variables? In both cases they should look the same for `TARGET==HOST` or not AFAIU.

In buildroot we use only the env varibles since we'd otherwise have to regenerate the config file for each package build due to our per-package build isolation changing things like linker paths for each package build, by using env variables we don't have to worry about that as the env is updated automatically with the correct paths for each package.

Additionally this makes it easier for us to handle say python package builds with cargo modules(ie python-cryptography) as we can simply append the cargo environment to our python environment without having to also trigger a config regeneration operation.